linkerd/stack/src/fail_on_error.rs

```diff
@@ -0,0 +1,88 @@
+use linkerd2_error::Error;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::sync::Mutex;
```

Comment:
AFAICT this lock is never held across an await, so it's probably more efficient to just use a `std::sync` mutex.